# Homework Assignment - Lesson 6: Combining Datasets - Joins

**Due Date:** [10/5/25]

**Instructions:**
- Complete the following tasks in this R notebook
- Use the exact variable names specified for each task
- Ensure your code runs without errors and produces the expected outputs
- Submit your completed notebook with all code and outputs

---

## Part 1: Data Import and Setup

**Task:** Import the following CSV files and examine their structure.

**Required Variable Names:**
You must use these EXACT variable names:
- `customers` for customers.csv
- `orders` for orders.csv  
- `order_items` for order_items.csv
- `products` for products.csv
- `suppliers` for suppliers.csv

**Expected Outputs:**
- All datasets loaded successfully
- Print dataset dimensions using `nrow()` and `ncol()`
- Show first 3 rows of each dataset using `head(dataset, 3)`

In [8]:
# Your code here:
# 1. Load the tidyverse package
# 2. Set working directory to sample_datasets  
# 3. Import all CSV files using read_csv() with EXACT variable names above
# 4. Print dimensions and head() for each dataset

# Load tidyverse
library(tidyverse)

# Set working directory
getwd()

# Import datasets with exact variable names
customers <- read_csv("customers.csv")
orders <- read_csv("orders.csv")
order_items <- read_csv("order_items.csv")
products <- read_csv("products.csv")
suppliers <- read_csv("suppliers.csv")

# Print dataset information (uncomment after creating variables above):
# cat("Dataset Dimensions:\n")
# cat("Customers:", nrow(customers), "rows x", ncol(customers), "columns\n")
# cat("Orders:", nrow(orders), "rows x", ncol(orders), "columns\n")
# cat("Order Items:", nrow(order_items), "rows x", ncol(order_items), "columns\n")
# cat("Products:", nrow(products), "rows x", ncol(products), "columns\n")
# cat("Suppliers:", nrow(suppliers), "rows x", ncol(suppliers), "columns\n")

# Show first 3 rows of each dataset (uncomment after loading data):
# head(customers, 3)
# head(orders, 3)
# head(order_items, 3)
# head(products, 3)
# head(suppliers, 3)

[1] "/workspaces/assignment-1-Chrstianfrrone/assignment/Homework"

Rows: 100 Columns: 5
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (3): Name, Email, City
dbl  (1): CustomerID
date (1): Registration_Date

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 250 Columns: 4


── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
dbl  (3): OrderID, CustomerID, Total_Amount
date (1): Order_Date

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 400 Columns: 4
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
dbl (4): OrderID, ProductID, Quantity, Unit_Price

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 50 Columns: 4
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (2): Product_Name, Category
dbl (2): ProductID, Supplier_ID

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 10 Columns: 3
── Colu

## Part 2: Basic Joins

**Tasks:**
1. **Inner Join:** Create `customer_orders_inner` by joining customers and orders
2. **Left Join:** Create `customer_orders_left` to show all customers  
3. **Right Join:** Create `customer_orders_right` to show all orders
4. **Full Join:** Create `customer_orders_full` for complete data view

Use the exact variable names specified above for autograding. Analyze row counts and explain results.

In [3]:
# Your code here:
# Create an inner join between customers and orders
# Analyze the results and compare row counts

customer_orders_inner <- customers %>%
    inner_join(orders, by = "CustomerID")
cat("Inner Join Results:\n")
cat("Customers:", nrow(customers), "\n")
cat("Orders:", nrow(orders), "\n")
cat("Inner Join Result:", nrow(customer_orders_inner), "\n")

Inner Join Results:
Customers: 100 
Orders: 250 
Inner Join Result: 200 


In [4]:
# Your code here:
# Create a left join to keep all customers
# Count how many customers have no orders

customer_orders_left <- customers %>%
  left_join(orders, by = "CustomerID")

customers_without_orders <- sum(is.na(customer_orders_left$OrderID))
cat("Left Join Results:\n")
cat("Total rows:", nrow(customer_orders_left), "\n")
cat("Customers without orders:", customers_without_orders, "\n")

Left Join Results:
Total rows: 200 
Customers without orders: 0 


In [5]:
# Your code here:
# Create a right join to keep all orders
# Check for orders with invalid customer IDs

customer_orders_right <- customers %>%
  right_join(orders, by = "CustomerID")

orders_invalid_customers <- sum(is.na(customer_orders_right$Name))
cat("Right Join Results:\n")
cat("Total rows:", nrow(customer_orders_right), "\n")
cat("Orders with invalid customer IDs:", orders_invalid_customers, "\n")

Right Join Results:
Total rows: 250 
Orders with invalid customer IDs: 50 


In [6]:
<VSCode.Cell id="#VSC-0da36c1e" language="markdown">
## Part 2: Basic Joins

**Required Variable Names:**
You must create these EXACT variable names:

**Part 2.1:** Create `customer_orders` using inner_join()
**Part 2.2:** Create `customer_orders_left` using left_join()  
**Part 2.3:** Create `customer_orders_right` using right_join()
**Part 2.4:** Create `customer_orders_full` using full_join()

**Expected Outputs:**
For each join, print:
- Number of rows in the result using `nrow()`
- Comparison with original dataset sizes
- Analysis of unmatched records

ERROR: Error in parse(text = input): <text>:1:1: unexpected '<'
1: <
    ^


In [9]:
# Part 2.1: Inner Join (REQUIRED variable name: customer_orders)
# Your code here:
customer_orders <- customers %>%
    inner_join(orders, by = "CustomerID")

# Required output for autograding:
cat("Inner Join Results:
")
cat("Customers:", nrow(customers), "
")
cat("Orders:", nrow(orders), "
")
cat("Inner Join Result:", nrow(customer_orders), "
")

Inner Join Results:
Customers: 100 
Orders: 250 
Inner Join Result: 200 


In [10]:
# Part 2.2: Left Join (REQUIRED variable name: customer_orders_left)  
# Your code here:
customer_orders_left <- customers %>%
  left_join(orders, by = "CustomerID")

# Required output for autograding:
customers_without_orders <- sum(is.na(customer_orders_left$OrderID))
cat("Left Join Results:
")
cat("Total rows:", nrow(customer_orders_left), "
")
cat("Customers without orders:", customers_without_orders, "
")

Left Join Results:
Total rows: 200 
Customers without orders: 0 


In [11]:
# Part 2.1: Inner Join (REQUIRED variable name: customer_orders)
# Your code here:
customer_oders <- customers %>%
    inner_join(orders, by = "CustomerID")

# Expected output:
cat("Inner Join Results:
")
cat("Customers:", nrow(customers), "
")
cat("Orders:", nrow(orders), "
")
cat("Inner Join Result:", nrow(customer_orders), "
")

Inner Join Results:
Customers: 100 
Orders: 250 
Inner Join Result: 200 


In [12]:
# Part 2.2: Left Join (REQUIRED variable name: customer_orders_left)  
# Your code here:
customers_rders_left <- customers %>%
    left_join(orders, by = "CustomerID")

# Expected output:
customers_without_orders <- sum(is.na(customer_orders_left$OrderID))
cat("Left Join Results:
")
cat("Total rows:", nrow(customer_orders_left), "
")
cat("Customers without orders:", customers_without_orders, "
")

Left Join Results:
Total rows: 200 
Customers without orders: 0 


In [13]:
# Part 2.3: Right Join (REQUIRED variable name: customer_orders_right)
# Your code here:
customer_orders_right <- customers%>%
  right_join(orders, by = "CustomerID")

# Expected output:
orders_invalid_customers <- sum(is.na(customer_orders_right$Name))
cat("Right Join Results:
")
cat("Total rows:", nrow(customer_orders_right), "
")
cat("Orders with invalid customer IDs:", orders_invalid_customers, "
")

Right Join Results:
Total rows: 250 
Orders with invalid customer IDs: 50 


In [14]:
# Part 2.4: Full Join (REQUIRED variable name: customer_orders_full)
# Your code here:
customer_orders_full <- customers %>%
    full_join(orders, by = "CustomerID")

# Expected output:
customers_only <- sum(is.na(customer_orders_full$OrderID))
orders_only <- sum(is.na(customer_orders_full$Name))
cat("Full Join Results:
")
cat("Total rows:", nrow(customer_orders_full), "
")
cat("Customers without orders:", customers_only, "
")
cat("Orders without valid customers:", orders_only, "
")

Full Join Results:
Total rows: 250 
Customers without orders: 0 
Orders without valid customers: 50 


In [15]:
# Part 2.4: Full Join (REQUIRED variable name: customer_orders_full)
# Your code here:
customer_orders_full <- customers %>%
    full_join(orders, by = "CustomerID")

# Required output for autograding:
customers_only <- sum(is.na(customer_orders_full$OrderID))
orders_only <- sum(is.na(customer_orders_full$Name))
cat("Full Join Results:
")
cat("Total rows:", nrow(customer_orders_full), "
")
cat("Customers without orders:", customers_only, "
")
cat("Orders without valid customers:", orders_only, "
")

Full Join Results:
Total rows: 250 
Customers without orders: 0 
Orders without valid customers: 50 


## Part 3: Multi-Table Joins

**Tasks:**
1. Create a comprehensive dataset by joining `orders`, `customers`, and `order_items`
2. Extend this dataset by adding `products` and `suppliers` information
3. Create a complete supply chain view showing the full customer-to-supplier relationship

Build your joins step by step and examine the results at each stage.

## Part 3: Multi-Table Joins

**Requirements for Autograding:**
You must create these EXACT variable names in order:

1. `orders_items` - Join orders and order_items  
2. `orders_customers_items` - Add customers to above
3. `complete_order_data` - Add products information
4. `complete_data` - Add suppliers for complete supply chain view

**Expected Outputs:**
Print the number of rows for each step to show the join progression.

In [16]:
# Step 1: REQUIRED variable name: orders_items
# Your code here:
orders_items <- orders %>%
    inner_join(order_items, by = "OrderID")

# Required output for autograding:
cat("Step 1 - Orders with Items:", nrow(orders_items), "rows\n")

Step 1 - Orders with Items: 400 rows


In [17]:
# Step 2: REQUIRED variable name: orders_customers_items  
# Your code here:
orders_customers_items <- orders_items %>%
    inner_join(customers, by = "CustomerID")

# Required output for autograding:
cat("Step 2 - Add Customers:", nrow(orders_customers_items), "rows\n")

Step 2 - Add Customers: 310 rows


In [18]:
# Step 3: REQUIRED variable name: complete_order_data
# Your code here:
complete_order_data <- orders_customers_items %>%
    inner_join(products, by = "ProductID")

# Required output for autograding:
cat("Step 3 - Add Products:", nrow(complete_order_data), "rows\n")

Step 3 - Add Products: 310 rows


In [19]:
# Step 4: REQUIRED variable name: complete_data
# Your code here:
complete_data <- complete_order_data %>%
  inner_join(suppliers, by = "Supplier_ID")


# Required output for autograding:
cat("Step 4 - Complete Supply Chain:", nrow(complete_data), "rows\n")

Step 4 - Complete Supply Chain: 310 rows


## Part 4: Data Quality Analysis

**Tasks:**
1. Use `anti_join()` to find unmatched records between tables
2. Use `semi_join()` to find matched records
3. Check for duplicate keys and analyze their impact on joins
4. Identify and document any data quality issues

This analysis helps understand data integrity and potential issues in your datasets.

In [ ]:
<VSCode.Cell id="#VSC-d7f91846" language="markdown">
## Part 4: Data Quality Analysis

**Requirements for Autograding:**
You must create these EXACT variable names:

1. `customers_no_orders` - Use anti_join() to find customers with no orders
2. `orphaned_orders` - Use anti_join() to find orders without customers  
3. `products_never_ordered` - Use anti_join() to find unordered products
4. `active_customers` - Use semi_join() to find customers with orders

**Expected Outputs:**
Print the count of records for each using `nrow()`.

ERROR: Error in parse(text = input): <text>:1:1: unexpected '<'
1: <
    ^


In [20]:
# REQUIRED variable name: customers_no_orders
# Your code here:
customers_no_orders <- customers %>%
    anti_join(orders, by = "CustomerID")

# Required output for autograding:
cat("Customers who never placed an order:", nrow(customers_no_orders), "\n")

Customers who never placed an order: 0 


In [21]:
# REQUIRED variable name: orphaned_orders
# Your code here:
orphaned_orders <- orders %>%
    anti_join(customers, by = "CustomerID")

# Required output for autograding:
cat("Orders without corresponding customers:", nrow(orphaned_orders), "\n")

Orders without corresponding customers: 50 


In [22]:
# REQUIRED variable name: products_never_ordered
# Your code here:
products_never_ordered <- products %>%
    anti_join(order_items, by = "ProductID")

# Required output for autograding:
cat("Products that were never ordered:", nrow(products_never_ordered), "\n")

Products that were never ordered: 0 


In [23]:
# REQUIRED variable name: active_customers
# Your code here:
active_customers <- customers %>%
    semi_join(orders, by = "CustomerID")

# Required output for autograding:
cat("Customers who placed at least one order:", nrow(active_customers), "\n")

Customers who placed at least one order: 100 


In [24]:
# Data integrity checks - count duplicate keys
# Your code here: Calculate and print duplicate counts for each dataset

# Required outputs for autograding (uncomment after creating the calculations above):
duplicate_customers <- sum(duplicated(customers$CustomerID))
duplicate_orders <- sum(duplicated(orders$OrderID))
duplicate_products <- sum(duplicated(products$ProductID))
duplicate_suppliers <- sum(duplicated(suppliers$Supplier_ID))

cat("Data Quality Summary:\n")
cat("Duplicate customer IDs:", duplicate_customers, "\n")
cat("Duplicate order IDs:", duplicate_orders, "\n")
cat("Duplicate product IDs:", duplicate_products, "\n")
cat("Duplicate supplier IDs:", duplicate_suppliers, "\n")

Data Quality Summary:
Duplicate customer IDs: 0 
Duplicate order IDs: 0 
Duplicate product IDs: 0 
Duplicate supplier IDs: 0 


## Part 5: Business Analysis with Joined Data

**Tasks:**
1. Calculate customer lifetime value metrics (total spent, number of orders, average order value)
2. Analyze product performance (total quantity sold, revenue generated, order frequency)
3. Evaluate supplier performance (total revenue, number of products supplied)
4. Create a regional analysis by customer city

Use your joined datasets to generate meaningful business insights.

In [ ]:
<VSCode.Cell id="#VSC-f7364f43" language="markdown">
## Part 5: Business Analysis with Joined Data

**Requirements for Autograding:**
You must create these EXACT variable names:

1. `customer_metrics` - Customer analysis with Total_Spent, Order_Count, Avg_Order_Value
2. `product_metrics` - Product analysis with Total_Revenue, Total_Quantity, Order_Frequency  
3. `supplier_metrics` - Supplier analysis with Product_Count by supplier
4. `regional_analysis` - Sales analysis by customer city

**Required Columns:**
- customer_metrics: CustomerID, Name, Total_Spent, Order_Count, Avg_Order_Value
- product_metrics: ProductID, Product_Name, Total_Revenue, Total_Quantity, Order_Frequency
- supplier_metrics: Supplier_ID, Supplier_Name, Product_Count
- regional_analysis: City, Total_Sales, Customer_Count, Avg_Customer_Value

ERROR: Error in parse(text = input): <text>:1:1: unexpected '<'
1: <
    ^


In [25]:
# REQUIRED variable name: customer_metrics
# Your code here: Calculate customer lifetime value metrics
# Required columns: CustomerID, Name, Total_Spent, Order_Count, Avg_Order_Value
customer_metrics <- orders_customers_items %>%
  dplyr::group_by(CustomerID, Name) %>%
  dplyr::summarise(
    Total_Spent    = sum(Quantity * Unit_Price, na.rm = TRUE),
    Order_Count    = dplyr::n_distinct(OrderID),
    Avg_Order_Value = Total_Spent / Order_Count,
    .groups = "drop"
  )

# Required output for autograding:
cat("Customer Analysis Summary:
")
cat("Total customers analyzed:", nrow(customer_metrics), "
")
cat("Top customer total spent: $", round(max(customer_metrics$Total_Spent, na.rm = TRUE), 2), "
")
print(head(customer_metrics[order(-customer_metrics$Total_Spent), ], 3))

Customer Analysis Summary:
Total customers analyzed: 94 
Top customer total spent: $ 8471.51 


# A tibble: 3 × 5
  CustomerID Name        Total_Spent Order_Count Avg_Order_Value
       <dbl> <chr>             <dbl>       <int>           <dbl>
1         53 Customer 53       8472.           2           4236.
2          3 Customer 3        6107.           2           3053.
3         61 Customer 61       5768.           2           2884.


In [26]:
# REQUIRED variable name: product_metrics  
# Your code here: Analyze product performance
# Required columns: ProductID, Product_Name, Total_Revenue, Total_Quantity, Order_Frequency
product_metrics <- complete_order_data %>%
  dplyr::group_by(ProductID, Product_Name) %>%
  dplyr::summarise(
    Total_Revenue   = sum(Quantity * Unit_Price, na.rm = TRUE),
    Total_Quantity  = sum(Quantity, na.rm = TRUE),
    Order_Frequency = dplyr::n_distinct(OrderID),
    .groups = "drop"
  ) %>%
  dplyr::arrange(dplyr::desc(Total_Revenue))

# Required output for autograding:
cat("Product Analysis Summary:
")
cat("Total products analyzed:", nrow(product_metrics), "
")
cat("Top product revenue: $", round(max(product_metrics$Total_Revenue, na.rm = TRUE), 2), "
")
print(head(product_metrics[order(-product_metrics$Total_Revenue), ], 3))

Product Analysis Summary:
Total products analyzed: 50 
Top product revenue: $ 11763.16 
# A tibble: 3 × 5
  ProductID Product_Name Total_Revenue Total_Quantity Order_Frequency
      <dbl> <chr>                <dbl>          <dbl>           <int>
1        47 Product 47          11763.             35               8
2        26 Product 26          10854.             42              13
3        43 Product 43          10584.             36              10


In [27]:
# REQUIRED variable name: supplier_metrics
# Your code here: Evaluate supplier performance  
# Required columns: Supplier_ID, Supplier_Name, Product_Count

supplier_metrics <- products %>%
  group_by(Supplier_ID) %>%
  summarise(Product_Count = n_distinct(ProductID), .groups = "drop") %>%
  left_join(suppliers %>% select(Supplier_ID, Supplier_Name), by = "Supplier_ID") %>%
  select(Supplier_ID, Supplier_Name, Product_Count)

# Required output for autograding:
cat("Supplier Analysis Summary:
")
cat("Total suppliers analyzed:", nrow(supplier_metrics), "
")
cat("Max products per supplier:", max(supplier_metrics$Product_Count, na.rm = TRUE), "
")
print(supplier_metrics[order(-supplier_metrics$Product_Count), ])

Supplier Analysis Summary:
Total suppliers analyzed: 10 
Max products per supplier: 10 
# A tibble: 10 × 3
   Supplier_ID Supplier_Name Product_Count
         <dbl> <chr>                 <int>
 1           7 Supplier 7               10
 2           4 Supplier 4                6
 3          10 Supplier 10               6
 4           1 Supplier 1                5
 5           3 Supplier 3                5
 6           5 Supplier 5                5
 7           6 Supplier 6                5
 8           2 Supplier 2                3
 9           8 Supplier 8                3
10           9 Supplier 9                2


In [28]:
# REQUIRED variable name: regional_analysis
# Your code here: Create regional analysis by customer city
# Required columns: City, Total_Sales, Customer_Count, Avg_Customer_Value
regional_analysis <- orders_customers_items %>%
  dplyr::group_by(City, CustomerID) %>%
  dplyr::summarise(Customer_Total = sum(Quantity * Unit_Price, na.rm = TRUE), .groups = "drop") %>%
  dplyr::group_by(City) %>%
  dplyr::summarise(
    Total_Sales        = sum(Customer_Total, na.rm = TRUE),
    Customer_Count     = dplyr::n_distinct(CustomerID),
    Avg_Customer_Value = ifelse(Customer_Count > 0, Total_Sales / Customer_Count, 0),
    .groups = "drop"
  )

# Required output for autograding:
cat("Regional Analysis Summary:
")
cat("Total cities analyzed:", nrow(regional_analysis), "
")
cat("Highest city sales: $", round(max(regional_analysis$Total_Sales, na.rm = TRUE), 2), "
")
print(regional_analysis[order(-regional_analysis$Total_Sales), ])

Regional Analysis Summary:
Total cities analyzed: 5 
Highest city sales: $ 72277.52 
# A tibble: 5 × 4
  City        Total_Sales Customer_Count Avg_Customer_Value
  <chr>             <dbl>          <int>              <dbl>
1 Chicago          72278.             23              3143.
2 Phoenix          49333.             19              2596.
3 Los Angeles      48506.             18              2695.
4 Houston          38143.             18              2119.
5 New York         36417.             16              2276.


In [29]:
# Your code here:
# Analyze product performance metrics
product_metrics <- complete_order_data %>%
  dplyr::group_by(ProductID, Product_Name) %>%
  dplyr::summarise(
    Total_Revenue   = sum(Quantity * Unit_Price, na.rm = TRUE),
    Total_Quantity  = sum(Quantity, na.rm = TRUE),
    Order_Frequency = dplyr::n_distinct(OrderID),
    .groups = "drop"
  )


In [30]:
# Your code here:
# Evaluate supplier performance
supplier_metrics <- products %>%
  group_by(Supplier_ID) %>%
  summarise(Product_Count = n_distinct(ProductID), .groups = "drop") %>%
  left_join(suppliers %>% select(Supplier_ID, Supplier_Name), by = "Supplier_ID") %>%
  select(Supplier_ID, Supplier_Name, Product_Count)


In [31]:
# Your code here:
# Create regional analysis by customer city
regional_analysis <- orders_customers_items %>%
    dplyr::group_by(City, CustomerID) %>%
    dplyr::summarise(Customer_Total = sum(Quantity * Unit_Price, na.rm = TRUE), .groups = "drop") %>%
    dplyr::group_by(City) %>%
    dplyr::summarise(
        Total_Sales        = sum(Customer_Total, na.rm = TRUE),
        Customer_Count     = dplyr::n_distinct(CustomerID),
        Avg_Customer_Value = ifelse(Customer_Count > 0, Total_Sales / Customer_Count, 0),
        .groups = "drop"
    )


In [32]:
# Supplier Performance Analysis
supplier_metrics <- complete_data %>%
  dplyr::group_by(Supplier_ID, Supplier_Name) %>%
  dplyr::summarise(
    Product_Count = dplyr::n_distinct(ProductID),
    .groups = "drop"
  ) %>%
  dplyr::arrange(dplyr::desc(Product_Count))

cat("Supplier Performance:\n")
print(supplier_metrics)

Supplier Performance:
# A tibble: 10 × 3
   Supplier_ID Supplier_Name Product_Count
         <dbl> <chr>                 <int>
 1           7 Supplier 7               10
 2           4 Supplier 4                6
 3          10 Supplier 10               6
 4           1 Supplier 1                5
 5           3 Supplier 3                5
 6           5 Supplier 5                5
 7           6 Supplier 6                5
 8           2 Supplier 2                3
 9           8 Supplier 8                3
10           9 Supplier 9                2


In [33]:
# Regional Analysis by Customer City
regional_analysis <- complete_order_data %>%
  group_by(City) %>%
  summarise(
    Total_Sales = sum(Quantity * Unit_Price, na.rm = TRUE),
    Customer_Count = n_distinct(CustomerID),
    Order_Count = n_distinct(OrderID),
    Avg_Customer_Value = Total_Sales / Customer_Count,
    .groups = "drop"
  ) %>%
  arrange(desc(Total_Sales))

cat("Regional Sales Analysis:\n")
print(regional_analysis)

Regional Sales Analysis:
# A tibble: 5 × 5
  City        Total_Sales Customer_Count Order_Count Avg_Customer_Value
  <chr>             <dbl>          <int>       <int>              <dbl>
1 Chicago          72278.             23          36              3143.
2 Phoenix          49333.             19          32              2596.
3 Los Angeles      48506.             18          32              2695.
4 Houston          38143.             18          30              2119.
5 New York         36417.             16          26              2276.


## Part 6: Complex Business Questions

**Tasks:**
Answer the following business questions using your joined datasets:

1. **Customer Segmentation:** Identify your top 10% of customers by total spending
2. **Product Recommendations:** Which products are frequently bought together?
3. **Supplier Dependency:** Which suppliers are most critical to the business?
4. **Market Expansion:** Which cities have high customer counts but low average order values?

Provide data-driven answers with supporting analysis.

## Part 6: Complex Business Questions

**Requirements for Autograding:**
You must create these EXACT variable names:

1. `top_customers` - Top 10% of customers by spending (use quantile function)
2. `product_combinations` - Products frequently bought together  
3. `critical_suppliers` - Suppliers ranked by importance
4. `market_expansion` - Cities with expansion opportunities

**Expected Outputs:**
Each analysis should include specific metrics and rankings as shown in the required outputs below.

In [34]:
# REQUIRED variable name: top_customers
# Your code here: Identify top 10% of customers by total spending
# Use quantile() function to find the 90th percentile threshold
top_10_percent_threshold <- quantile(customer_metrics$Total_Spent, 0.9, na.rm = TRUE)

top_customers <- customer_metrics %>%
    filter(Total_Spent >= top_10_percent_threshold) %>%
    arrange(dplyr::desc(Total_Spent))

cat("Customer Segmentation Results:\n")
cat("Top 10% spending threshold: $", round(top_10_percent_threshold, 2), "\n")
cat("Number of top customers:", nrow(top_customers), "\n")
print(head(top_customers, 5))
# Required output for autograding (uncomment after completing your analysis):
top_10_percent_threshold <- quantile(customer_metrics$Total_Spent, 0.9, na.rm = TRUE)
cat("Customer Segmentation Results:\n")
cat("Top 10% spending threshold: $", round(top_10_percent_threshold, 2), "\n")
cat("Number of top customers:", nrow(top_customers), "\n")

Customer Segmentation Results:
Top 10% spending threshold: $ 4931.51 
Number of top customers: 10 
# A tibble: 5 × 5
  CustomerID Name        Total_Spent Order_Count Avg_Order_Value
       <dbl> <chr>             <dbl>       <int>           <dbl>
1         53 Customer 53       8472.           2           4236.
2          3 Customer 3        6107.           2           3053.
3         61 Customer 61       5768.           2           2884.
4         82 Customer 82       5746.           2           2873.
5         80 Customer 80       5714.           2           2857.
Customer Segmentation Results:
Top 10% spending threshold: $ 4931.51 
Number of top customers: 10 


In [35]:
# REQUIRED variable name: product_combinations
# Your code here: Find products frequently bought together
# Hint: Group by OrderID, filter orders with multiple items
product_combinations <- order_items %>%
  select(OrderID, ProductID) %>%
  distinct() %>%                        
  inner_join(., ., by = "OrderID") %>%  
  filter(ProductID.x < ProductID.y) %>% 
  count(ProductID.x, ProductID.y, name = "n") %>%
  arrange(desc(n))
cat("Product Combination Analysis:\n")


cat("Product Combination Analysis:\n")
cat("Total product combinations found:", nrow(product_combinations), "\n")
if(nrow(product_combinations) > 0) {
cat("Most frequent combination appears", max(product_combinations$n, na.rm = TRUE), "times\n")
}

Warning message in inner_join(., ., by = "OrderID"):
“Detected an unexpected many-to-many relationship between `x` and `y`.
ℹ Row 2 of `x` matches multiple rows in `y`.
ℹ Row 5 of `y` matches multiple rows in `x`.
ℹ If a many-to-many relationship is expected, set `relationship =
  "many-to-many"` to silence this warning.”


Product Combination Analysis:
Product Combination Analysis:
Total product combinations found: 281 
Most frequent combination appears 3 times


In [36]:
# REQUIRED variable name: critical_suppliers
# Your code here: Analyze which suppliers are most critical
# Consider both revenue contribution and product diversity
df <- complete_data
if ("SupplierID" %in% names(df)) df <- dplyr::rename(df, Supplier_ID = SupplierID)
if ("Name" %in% names(df) && !"Supplier_Name" %in% names(df)) {
  df <- dplyr::rename(df, Supplier_Name = Name)
}

critical_suppliers <- df %>%
  dplyr::group_by(Supplier_ID, Supplier_Name) %>%
  dplyr::summarise(
    Total_Revenue   = sum(Quantity * Unit_Price, na.rm = TRUE),
    Product_Count   = dplyr::n_distinct(ProductID),
    Orders_Involved = dplyr::n_distinct(OrderID),
    .groups = "drop"
  ) %>%
  dplyr::arrange(dplyr::desc(Total_Revenue), dplyr::desc(Product_Count))

cat("Critical Suppliers Analysis:\n")
cat("Total suppliers analyzed:", nrow(critical_suppliers), "\n")
cat("Top 3 most critical suppliers:\n")
print(head(critical_suppliers, 3))

Critical Suppliers Analysis:
Total suppliers analyzed: 10 
Top 3 most critical suppliers:
# A tibble: 3 × 5
  Supplier_ID Supplier_Name Total_Revenue Product_Count Orders_Involved
        <dbl> <chr>                 <dbl>         <int>           <int>
1           5 Supplier 5           43434.             5              43
2           7 Supplier 7           35416.            10              48
3          10 Supplier 10          31875.             6              35


In [37]:
# REQUIRED variable name: market_expansion  
# Your code here: Identify market expansion opportunities
# Focus on cities with multiple customers but varying order values
market_expansion <- orders_customers_items %>%
  dplyr::group_by(City) %>%
  dplyr::summarise(
    Total_Sales        = sum(Quantity * Unit_Price, na.rm = TRUE),
    Customer_Count     = dplyr::n_distinct(CustomerID),
    Order_Count        = dplyr::n_distinct(OrderID),
    Avg_Order_Value    = ifelse(Order_Count > 0, Total_Sales / Order_Count, 0),
    Avg_Customer_Value = ifelse(Customer_Count > 0, Total_Sales / Customer_Count, 0),
    .groups = "drop"
  ) %>%

dplyr::filter(Customer_Count >= 3) %>%
  dplyr::arrange(dplyr::desc(Customer_Count), Avg_Order_Value)
# Required output for autograding (uncomment after completing your analysis):
cat("Market Expansion Analysis:\n")
cat("Cities evaluated for expansion:", nrow(market_expansion), "\n")
print("Top expansion opportunities:")
print(head(market_expansion, 5))

Market Expansion Analysis:
Cities evaluated for expansion: 5 
[1] "Top expansion opportunities:"
# A tibble: 5 × 6
  City        Total_Sales Customer_Count Order_Count Avg_Order_Value
  <chr>             <dbl>          <int>       <int>           <dbl>
1 Chicago          72278.             23          36           2008.
2 Phoenix          49333.             19          32           1542.
3 Houston          38143.             18          30           1271.
4 Los Angeles      48506.             18          32           1516.
5 New York         36417.             16          26           1401.
# ℹ 1 more variable: Avg_Customer_Value <dbl>


## Part 7: Summary and Insights

**Requirements for Autograding:**
Complete the analysis summary below with specific findings from your work. Your responses will be evaluated for completeness and accuracy.

**Grading Criteria:**
- Specific metrics and numbers from your analysis
- Clear identification of data quality issues  
- Actionable business recommendations
- Technical insights about join performance and appropriateness

## Analysis Summary

### Key Findings from Join Operations:
- **Join Efficiency:** [Orders with valid customers: 80%. Customers without orders: 0%. Orders with invalid/missing customer IDs: 50. Full join unmatched: customers_only = 0, orders_only = 50]
- **Data Relationships:** [Customers to orders and orders to items are both one-to-many. Items to product and products to suppliers are many-to-one.]
- **Multi-table Joins:** [The row count expanded only once for orders to items. After that, every join was 1:1 and full matched so the row count stayed at 310 through the products and suppliers.]

### Data Quality Issues Discovered:
- **Orphaned Records:** [Orders without customers 50. Customers without orders: 0.]
- **Missing Data:** [The 50 orders without customers are due to missing or invalid customer IDs in the orders table.]
- **Referential Integrity:** [FK violation on orders.customer_id referencing customers.customer_id due to missing/invalid IDs.]

### Business Insights:
- **Top Customer:** [Customer 53 spent $8472]
- **Best Product:** [Product 47 had $11763 in revenue]
- **Regional Performance:** [Chicago: 23 customers, 36 orders, average order value = $2,008. Phoenix: 19 customers, 32 orders, average order value = $1,542. Los Angeles: 18 customers, 32 orders, average order value = $1,516. New York: 16 customers, 26 orders, average order value = $1,401]
- **Supplier Analysis:** [Top revenue suppliers: Supplier 5 ($43,434), Supplier 7 ($35,416), Supplier 10 ($31,875).]

### Strategic Recommendations:
- [1. Fix the bad orders, 50 orders are missing a real customer. 2. Treat top customers with perks or loyalty programs to retain loyalty. 3. Bundle frequently or top products with lower performing products to boost their sales.]

### Technical Learnings:
- **Inner vs Left Joins:** [I used inner joins when I wanted only valid matches and left joins when I wanted to keep all the records on the left side.]
- **Multi-table Strategy:** [1. Orders to order items. 2. add customers. 3. add products. 4. add suppliers.]
- **Performance Considerations:** [Filter early, select only needed columns before joins. Cut the data early for faster joins. Remove duplicates if any. Summarize after grouping not per row.]

In [ ]:
# Optional: Save your analysis results (not required for autograding)
# Your code here: Save key datasets to CSV files if desired


# Final summary output for autograding verification:
cat("\n=== HOMEWORK 6 COMPLETION SUMMARY ===\n")
cat("✓ Data Import: 5 datasets loaded\n")
cat("✓ Basic Joins: 4 join types completed\n") 
cat("✓ Multi-table Joins: 4-step progression completed\n")
cat("✓ Data Quality: Anti-joins and semi-joins performed\n")
cat("✓ Business Analysis: Customer, product, supplier, and regional analysis completed\n")
cat("✓ Complex Questions: Advanced business scenarios analyzed\n")
cat("✓ Summary: Comprehensive insights documented\n")
cat("\nAll required variables created for autograding verification.\n")